include/electionguard/precompute_buffers.hpp

```diff
@@ -112,32 +106,178 @@ namespace electionguard
     /// </summary>
     class EG_API TwoTriplesAndAQuadruple
     {
-        unique_ptr<Triple> triple1;
-        unique_ptr<Triple> triple2;
-        unique_ptr<Quadruple> quad;
+        std::unique_ptr<Triple> triple1;
+        std::unique_ptr<Triple> triple2;
+        std::unique_ptr<Quadruple> quad;
 
       public:
         explicit TwoTriplesAndAQuadruple() {}
-        TwoTriplesAndAQuadruple(unique_ptr<Triple> in_triple1, unique_ptr<Triple> in_triple2,
-                                unique_ptr<Quadruple> in_quad);
+        TwoTriplesAndAQuadruple(std::unique_ptr<Triple> in_triple1,
+                                std::unique_ptr<Triple> in_triple2,
+                                std::unique_ptr<Quadruple> in_quad);
         TwoTriplesAndAQuadruple(const TwoTriplesAndAQuadruple &other);
         TwoTriplesAndAQuadruple(TwoTriplesAndAQuadruple &&);
         ~TwoTriplesAndAQuadruple();
 
         TwoTriplesAndAQuadruple &operator=(const TwoTriplesAndAQuadruple &other);
         TwoTriplesAndAQuadruple &operator=(TwoTriplesAndAQuadruple &&);
 
-        unique_ptr<Triple> get_triple1() { return triple1->clone(); }
+        std::unique_ptr<Triple> get_triple1() { return triple1->clone(); }
+
+        std::unique_ptr<Triple> get_triple2() { return triple2->clone(); }
+
+        std::unique_ptr<Quadruple> get_quad() { return quad->clone(); }
+
+        std::unique_ptr<TwoTriplesAndAQuadruple> clone();
+    };
+
+    /// <summary>
+    /// A buffer of precomputed values that are used to speed up encryption
+    /// of a selection. Since the values are precomputed it removes many the
+    /// exponentiations from the ElGamal encryption of the selection as well
+    /// as the computation of the Chaum Pedersen proof.
+    ///
+    /// The precompute buffer is a queue of TwoTriplesAndAQuadruple objects.
+    /// The queue is filled by a background thread. The background thread
+    /// will fill the queue until it reaches the max queue size. The max
+    /// queue size is set by the caller and defaults to 5000. The queue
+    /// size is set by the caller in the init method.
+    ///
+    /// This class is initialized against a specific public key and is thread safe.
+    /// </summary>
+    class EG_API PrecomputeBuffer
```

Comment:
Precompute buffer is now broken out into a a standard class and a static "context". I needed to do this to better define the scope of the class data and make it easier to debug. Now the behavior is consistent and we could support having more than one precompute buffer (e.g. one per election key) if we wanted.